// @prismicio/api-renderer/lib/widgets/GroupRenderer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const SimpleWidgetRenderer_1 = tslib_1.__importDefault(require("./SimpleWidgetRenderer"));
const GroupRenderer = (ctx) => ({
    renderV1(content, fetch) {
        return {
            type: "Group",
            value: content.value.map((items) => {
                return (0, SimpleWidgetRenderer_1.default)(ctx).renderObjectOfSimpleWidgetV1(items.value, fetch === null || fetch === void 0 ? void 0 : fetch.fields);
            }),
        };
    },
    renderV2(groupDef, content, fetch) {
        return content.value.map((items) => (0, SimpleWidgetRenderer_1.default)(ctx).renderObjectOfSimpleWidgetV2(items.value, fetch === null || fetch === void 0 ? void 0 : fetch.fields, groupDef.fields, {
            withFetch: "filter",
        }, {
            default: "default",
            withFetch: "default",
            withFetchField: "default",
        }));
    },
    renderDefault(_) {
        return [];
    },
    renderMocks(def, content) {
        return content.value.map((items) => (0, SimpleWidgetRenderer_1.default)(ctx).renderObjectOfSimpleWidgetMocks(items.value, def.fields));
    },
});
exports.default = GroupRenderer;
