// @slicemachine/manager/dist/managers/telemetry/types.js
const SegmentEventType = {
  command_init_start: "command:init:start",
  command_init_identify: "command:init:identify",
  command_init_end: "command:init:end",
  review: "review",
  sliceSimulator_setup: "slice-simulator:setup",
  sliceSimulator_open: "slice-simulator:open",
  sliceSimulator_isNotRunning: "slice-simulator:is-not-running",
  pageView: "page-view",
  openVideoTutorials: "open-video-tutorials",
  customType_created: "custom-type:created",
  customType_fieldAdded: "custom-type:field-added",
  customType_sliceZoneUpdated: "custom-type:slice-zone-updated",
  customType_openAddFromTemplates: "custom-type:open-add-from-templates",
  customType_saved: "custom-type:saved",
  slice_created: "slice:created",
  screenshotTaken: "screenshot-taken",
  changes_pushed: "changes:pushed",
  changes_limitReach: "changes:limit-reach",
  editor_widgetUsed: "editor:widget-used",
  open_page_snippet: "page-type:open-snippet",
  copy_page_snippet: "page-type:copy-snippet"
};
const HumanSegmentEventType = {
  [SegmentEventType.command_init_start]: "SliceMachine Init Start",
  [SegmentEventType.command_init_identify]: "SliceMachine Init Identify",
  [SegmentEventType.command_init_end]: "SliceMachine Init End",
  [SegmentEventType.review]: "SliceMachine Review",
  [SegmentEventType.sliceSimulator_setup]: "SliceMachine Slice Simulator Setup",
  [SegmentEventType.sliceSimulator_open]: "SliceMachine Slice Simulator Open",
  [SegmentEventType.sliceSimulator_isNotRunning]: "SliceMachine Slice Simulator is not running",
  [SegmentEventType.pageView]: "SliceMachine Page View",
  [SegmentEventType.openVideoTutorials]: "SliceMachine Open Video Tutorials",
  [SegmentEventType.customType_created]: "SliceMachine Custom Type Created",
  [SegmentEventType.customType_fieldAdded]: "SliceMachine Custom Type Field Added",
  [SegmentEventType.customType_sliceZoneUpdated]: "SliceMachine Slicezone Updated",
  [SegmentEventType.customType_openAddFromTemplates]: "SliceMachine Open Add from templates",
  [SegmentEventType.customType_saved]: "SliceMachine Custom Type Saved",
  [SegmentEventType.slice_created]: "SliceMachine Slice Created",
  [SegmentEventType.screenshotTaken]: "SliceMachine Screenshot Taken",
  [SegmentEventType.changes_pushed]: "SliceMachine Changes Pushed",
  [SegmentEventType.changes_limitReach]: "SliceMachine Changes Limit Reach",
  [SegmentEventType.editor_widgetUsed]: "SliceMachine Editor Widget Used",
  [SegmentEventType.open_page_snippet]: "SliceMachine Opens Page Type Snippet Dialog",
  [SegmentEventType.copy_page_snippet]: "Slice Machine page code snippet copied"
};
export {
  HumanSegmentEventType,
  SegmentEventType
};
