// @slicemachine/manager/dist/constants/API_ENDPOINTS.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const trailingSlash = require("../lib/trailingSlash.cjs");
const APPLICATION_MODE = require("./APPLICATION_MODE.cjs");
const API_ENDPOINTS = (() => {
  switch (process.env.SM_ENV) {
    case APPLICATION_MODE.APPLICATION_MODE.Development: {
      const apiEndpoints = {
        PrismicWroom: trailingSlash.addTrailingSlash(process.env.wroom_endpoint),
        PrismicAuthentication: trailingSlash.addTrailingSlash(process.env.authentication_server_endpoint),
        PrismicModels: trailingSlash.addTrailingSlash(process.env.customtypesapi_endpoint),
        PrismicUser: trailingSlash.addTrailingSlash(process.env.user_service_endpoint),
        AwsAclProvider: trailingSlash.addTrailingSlash(process.env.acl_provider_endpoint),
        PrismicOembed: trailingSlash.removeTrailingSlash(process.env.oembed_endpoint ?? "https://oembed.wroom.io"),
        PrismicUnsplash: trailingSlash.removeTrailingSlash(process.env.unsplash_endpoint ?? "https://unsplash.wroom.io")
      };
      const missingAPIEndpoints = Object.keys(apiEndpoints).filter((key) => {
        return !apiEndpoints[key];
      });
      if (missingAPIEndpoints.length > 0) {
        console.error(`You are running Slice Machine in development mode where API endpoints are configured via environment variables. The following endpoints were not configured: ${missingAPIEndpoints.join(", ")}. Configure them before continuing.`);
        process.exit(1);
      }
      return apiEndpoints;
    }
    case APPLICATION_MODE.APPLICATION_MODE.Staging: {
      return {
        PrismicWroom: "https://wroom.io/",
        PrismicAuthentication: "https://auth.wroom.io/",
        PrismicModels: "https://customtypes.wroom.io/",
        PrismicUser: "https://user.wroom.io/",
        AwsAclProvider: "https://2iamcvnxf4.execute-api.us-east-1.amazonaws.com/stage/",
        PrismicOembed: "https://oembed.wroom.io",
        PrismicUnsplash: "https://unsplash.wroom.io"
      };
    }
    case APPLICATION_MODE.APPLICATION_MODE.Production:
    default: {
      return {
        PrismicWroom: "https://prismic.io/",
        PrismicAuthentication: "https://auth.prismic.io/",
        PrismicModels: "https://customtypes.prismic.io/",
        PrismicUser: "https://user.internal-prismic.io/",
        AwsAclProvider: "https://0yyeb2g040.execute-api.us-east-1.amazonaws.com/prod/",
        PrismicOembed: "https://oembed.prismic.io",
        PrismicUnsplash: "https://unsplash.prismic.io"
      };
    }
  }
})();
exports.API_ENDPOINTS = API_ENDPOINTS;
