// @prismicio/types-internal/lib/customtypes/widgets/slices/CompositeSlice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const t = (0, tslib_1.__importStar)(require("io-ts"));
const widgets_1 = require("../../../documents/widgets");
const validators_1 = require("../../../validators");
const NestableWidget_1 = (0, tslib_1.__importDefault)(require("../nestable/NestableWidget"));
const SlicesTypes_1 = require("./SlicesTypes");
const CompositeSliceConfig = t.exact(t.partial({
    label: validators_1.StringOrNull,
}));
const CompositeSlice = t.exact(t.intersection([
    t.type({
        type: t.literal(SlicesTypes_1.SlicesTypes.Slice),
    }),
    t.partial({
        fieldset: validators_1.StringOrNull,
        description: t.string,
        icon: t.string,
        display: t.string,
        "non-repeat": t.record(widgets_1.WidgetKey, NestableWidget_1.default),
        repeat: t.record(widgets_1.WidgetKey, NestableWidget_1.default),
        config: CompositeSliceConfig,
    }),
]));
exports.default = CompositeSlice;
