// @prismicio/types-internal/lib/customtypes/widgets/slices/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SlicesTypes = exports.SliceZone = exports.Slice = exports.SharedSliceRef = exports.LegacySlice = exports.CompositeSlice = void 0;
const tslib_1 = require("tslib");
var CompositeSlice_1 = require("./CompositeSlice");
Object.defineProperty(exports, "CompositeSlice", { enumerable: true, get: function () { return (0, tslib_1.__importDefault)(CompositeSlice_1).default; } });
var LegacySlice_1 = require("./LegacySlice");
Object.defineProperty(exports, "LegacySlice", { enumerable: true, get: function () { return (0, tslib_1.__importDefault)(LegacySlice_1).default; } });
(0, tslib_1.__exportStar)(require("./SharedSlice"), exports);
var SharedSliceRef_1 = require("./SharedSliceRef");
Object.defineProperty(exports, "SharedSliceRef", { enumerable: true, get: function () { return (0, tslib_1.__importDefault)(SharedSliceRef_1).default; } });
exports.Slice = (0, tslib_1.__importStar)(require("./Slice"));
exports.SliceZone = (0, tslib_1.__importStar)(require("./Slices"));
var SlicesTypes_1 = require("./SlicesTypes");
Object.defineProperty(exports, "SlicesTypes", { enumerable: true, get: function () { return SlicesTypes_1.SlicesTypes; } });
