// @prismicio/api-renderer/src/widgets/slices/SharedSliceRenderer.ts
import type { SharedSliceContent } from "@prismicio/types-internal/lib/documents/widgets/slices"

import type { RenderContext, SharedSliceDefO, VariationDef } from "../../models"
import type { Field, SharedSlice } from "../../models/fetch"
import type { SliceRenderer } from "../../models/Renderer"
import SimpleWidgetRenderer from "../SimpleWidgetRenderer"

function getFetchField(
	variation: string,
	part: "repeat" | "nonRepeat",
	fetch: SharedSlice | undefined,
): Record<string, Field> | undefined {
	return fetch?.variations?.[variation]?.[part]?.fields
}

const SharedSliceRenderer: (ctx: RenderContext) => SliceRenderer<SharedSliceDefO, SharedSliceContent, SharedSlice> = (
	ctx,
) => ({
	renderV1(content: SharedSliceContent, fetch: SharedSlice | undefined): unknown {
		const itemsValue = content.items.map((groupItem) => {
			return SimpleWidgetRenderer(ctx).renderObjectOfSimpleWidgetV1(
				groupItem.value,
				getFetchField(content.variation, "repeat", fetch),
			)
		})
		const primaryValue = SimpleWidgetRenderer(ctx).renderObjectOfSimpleWidgetV1(
			content.primary,
			getFetchField(content.variation, "nonRepeat", fetch),
		)
		return {
			variation: content.variation,
			items: itemsValue,
			primary: primaryValue,
		}
	},

	renderV2(def: SharedSliceDefO, content: SharedSliceContent, fetch: SharedSlice | undefined): unknown {
		const variation = def.variations.find((variation: VariationDef) => variation.id === content.variation)
		if (variation === undefined) {
			return this.renderDefault(def)
		}

		const itemsValue = content.items.map((groupItem) => {
			return SimpleWidgetRenderer(ctx).renderObjectOfSimpleWidgetV2(
				groupItem.value,
				getFetchField(content.variation, "repeat", fetch),
				variation.items,
				{
					withFetch: "filter",
				},
				{
					default: "default",
					withFetch: "filter",
					withFetchField: "render",
				},
			)
		})
		const primaryValue = SimpleWidgetRenderer(ctx).renderObjectOfSimpleWidgetV2(
			content.primary,
			getFetchField(content.variation, "nonRepeat", fetch),
			variation.primary,
			{
				withFetch: "filter",
			},
			{
				default: "default",
				withFetch: "filter",
				withFetchField: "render",
			},
		)

		return {
			variation: content.variation,
			version: variation.version,
			items: itemsValue,
			primary: primaryValue,
		}
	},

	renderDefault(_def: SharedSliceDefO): unknown {
		return {}
	},

	renderMocks(def: SharedSliceDefO, content: SharedSliceContent): unknown {
		const variation = def.variations.find((variation: VariationDef) => variation.id === content.variation)
		if (variation === undefined) {
			return this.renderDefault(def)
		}

		const itemsValue = content.items.map((groupItem) => {
			return SimpleWidgetRenderer(ctx).renderObjectOfSimpleWidgetMocks(groupItem.value, variation.items)
		})
		const primaryValue = SimpleWidgetRenderer(ctx).renderObjectOfSimpleWidgetMocks(content.primary, variation.primary)

		return {
			variation: content.variation,
			version: variation.version,
			items: itemsValue,
			primary: primaryValue,
		}
	},
})

export default SharedSliceRenderer
