// @prismicio/types-internal/src/content/fields/nestable/NestableContent.ts
import * as t from "io-ts"

import {
	EmbedContentType,
	StructuredTextContentType,
} from "../../../documents/widgets/nestable"
import { BooleanContent, BooleanContentType } from "./BooleanContent"
import { EmbedContent } from "./EmbedContent"
import { FieldContent, FieldContentType } from "./FieldContent"
import { GeoPointContent, GeoPointContentType } from "./GeoPointContent"
import { ImageContent, ImageContentType } from "./ImageContent"
import {
	IntegrationFieldContent,
	IntegrationFieldsContentType,
} from "./IntegrationFieldContent"
import { LinkContent, LinksTypes } from "./LinkContent"
import { RichTextContent } from "./RichTextContent"
import { SeparatorContent, SeparatorContentType } from "./SeparatorContent"

export const NestableContent = t.union([
	BooleanContent,
	EmbedContent,
	FieldContent,
	GeoPointContent,
	ImageContent,
	IntegrationFieldContent,
	LinkContent,
	RichTextContent,
	SeparatorContent,
])

export type NestableContent = t.TypeOf<typeof NestableContent>

export const NestableTypes = {
	Links: LinksTypes,
	RichText: StructuredTextContentType,
	Boolean: BooleanContentType,
	Embed: EmbedContentType,
	Field: FieldContentType,
	GeoPoint: GeoPointContentType,
	Image: ImageContentType,
	IntegrationFields: IntegrationFieldsContentType,
	Separator: SeparatorContentType,
} as const
