// @prismicio/types-internal/lib/customtypes/widgets/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WidgetTypes = exports.Widget = exports.UID = exports.Slices = exports.Shared = exports.Nestable = exports.Group = void 0;
const tslib_1 = require("tslib");
var Group_1 = require("./Group");
Object.defineProperty(exports, "Group", { enumerable: true, get: function () { return (0, tslib_1.__importDefault)(Group_1).default; } });
exports.Nestable = (0, tslib_1.__importStar)(require("./nestable"));
exports.Shared = (0, tslib_1.__importStar)(require("./shared"));
exports.Slices = (0, tslib_1.__importStar)(require("./slices"));
var UID_1 = require("./UID");
Object.defineProperty(exports, "UID", { enumerable: true, get: function () { return (0, tslib_1.__importDefault)(UID_1).default; } });
exports.Widget = (0, tslib_1.__importStar)(require("./Widget"));
var WidgetTypes_1 = require("./WidgetTypes");
Object.defineProperty(exports, "WidgetTypes", { enumerable: true, get: function () { return WidgetTypes_1.WidgetTypes; } });
