// @prismicio/types-internal/src/documents/widgets/index.ts
import * as t from "io-ts"

import { refineType } from "../../validators/function"
import { EmptyContentType } from "./EmptyContent"
import { GroupContentType } from "./GroupContent"
import { NestableTypes } from "./nestable"
import { SlicesContent, SlicesContentType } from "./slices"
import StaticWidgetContent from "./StaticWidgetContent"
import { UIDContentType } from "./UIDContent"

export * from "./EmptyContent"
export * from "./GroupContent"
export * as Nestable from "./nestable"
export * from "./SimpleWidgetContent"
export * as Slices from "./slices"
export {
	type NonEmptyStaticWidgetContent,
	default as StaticWidgetContent,
} from "./StaticWidgetContent"
export * from "./UIDContent"

export const WidgetTypes = {
	...NestableTypes,
	Empty: EmptyContentType,
	UID: UIDContentType,
	Group: GroupContentType,
	slices: SlicesContentType,
} as const

export type WidgetTypes = typeof WidgetTypes[keyof typeof WidgetTypes]

const widgetKeyRegex = new RegExp("^[^<>]+$")

export const WidgetKey = refineType(
	t.string,
	"WidgetKey",
	(s: string) => s.length === 0 || widgetKeyRegex.test(s),
)

export type WidgetKey = t.TypeOf<typeof WidgetKey>

export type WidgetContent = SlicesContent | StaticWidgetContent

export const Widget = {
	fromJson(
		widgetKey: WidgetKey,
		widgetValue: unknown,
		widgetTypes: Map<WidgetKey, string>,
		widgetPositions: Map<WidgetKey, number>,
	) {
		const fieldType = widgetTypes.get(widgetKey)

		if (fieldType === "Slices") {
			return SlicesContent.fromJson(
				widgetKey,
				widgetValue,
				widgetTypes,
				widgetPositions,
			)
		} else {
			return StaticWidgetContent.fromJson(
				[],
				widgetKey,
				widgetValue,
				widgetTypes,
				widgetPositions,
			)
		}
	},
}
