// @prismicio/types-internal/src/documents/widgets/nestable/ImageContent.ts
import * as t from "io-ts"

import { NonEmptyStringOrNull, StringOrNull } from "../../../validators"
import { addType } from "../../../validators/function"

const Origin = t.exact(
	t.type({
		id: t.string,
		url: t.string,
		width: t.number,
		height: t.number,
	}),
)

export const ImageContentView = t.exact(
	t.intersection([
		t.type({
			origin: Origin,
			width: t.number,
			height: t.number,
			edit: t.type({
				zoom: t.number,
				crop: t.type({
					x: t.number,
					y: t.number,
				}),
				background: t.string,
			}),
		}),
		t.partial({
			url: t.string,
			credits: NonEmptyStringOrNull,
			alt: NonEmptyStringOrNull,
			provider: StringOrNull,
		}),
	]),
)
export type ImageContentView = t.TypeOf<typeof ImageContentView>

export const ImageContentType = "ImageContent"
export const ImageContent = addType(
	t.intersection([
		ImageContentView,
		t.partial({
			thumbnails: t.record(t.string, ImageContentView),
		}),
	]),
	ImageContentType,
)
export type ImageContent = t.TypeOf<typeof ImageContent>
export type ImageContentO = t.OutputOf<typeof ImageContent>
