// @prismicio/types-internal/src/documents/widgets/nestable/index.ts
import { BooleanContentType } from "./BooleanContent"
import { EmbedContentType } from "./EmbedContent"
import { FieldContentType } from "./FieldContent"
import { GeoPointContentType } from "./GeoPointContent"
import { ImageContentType } from "./ImageContent"
import { IntegrationFieldsContentType } from "./IntegrationFieldsContent"
import { LinksTypes } from "./Link"
import { SeparatorContentType } from "./SeparatorContent"
import { StructuredTextContentType } from "./StructuredTextContent"

export { default as BooleanContent } from "./BooleanContent"
export * from "./EmbedContent"
export { default as FieldContent } from "./FieldContent"
export { type GeoPointContentO, GeoPointContent } from "./GeoPointContent"
export {
	type ImageContentO,
	ImageContent,
	ImageContentView,
} from "./ImageContent"
export { default as IntegrationFieldsContent } from "./IntegrationFieldsContent"
export * as Links from "./Link"
export * from "./SeparatorContent"
export * from "./StructuredTextContent"

export const NestableTypes = {
	Links: LinksTypes,
	StructuredText: StructuredTextContentType,
	Boolean: BooleanContentType,
	Embed: EmbedContentType,
	Field: FieldContentType,
	GeoPoint: GeoPointContentType,
	Image: ImageContentType,
	IntegrationFields: IntegrationFieldsContentType,
	Separator: SeparatorContentType,
} as const
