// @prismicio/api-renderer/lib/models/ApiModel/Link/DocumentLink.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultDocumentLinkDef = exports.fromDocumentLink = exports.DocumentLinkDef = exports.DocumentLinkDefType = void 0;
const tslib_1 = require("tslib");
const function_1 = require("@prismicio/types-internal/lib/validators/function");
const t = tslib_1.__importStar(require("io-ts"));
exports.DocumentLinkDefType = "ConfigDocumentLinkDef";
exports.DocumentLinkDef = (0, function_1.addType)(t.strict({
    select: t.literal("document"),
    customtypes: t.readonlyArray(t.string),
}), exports.DocumentLinkDefType);
function fromDocumentLink(link) {
    var _a;
    if (!("document" === ((_a = link.config) === null || _a === void 0 ? void 0 : _a.select)))
        throw new Error('Invalid link type provided, should contain { select: "document" }');
    return exports.DocumentLinkDef.encode({
        select: "document",
        customtypes: link.config.customtypes || [],
    });
}
exports.fromDocumentLink = fromDocumentLink;
function defaultDocumentLinkDef() {
    return exports.DocumentLinkDef.encode({
        select: "document",
        customtypes: [],
    });
}
exports.defaultDocumentLinkDef = defaultDocumentLinkDef;
