// @prismicio/api-renderer/lib/models/ApiModel/Link/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isWebLinkDef = exports.isDocumentLinkDef = exports.isAnyLinkDef = exports.isMediaLinkDef = exports.defaultLinkDef = exports.fromLink = exports.LinkDef = exports.LinkDefType = void 0;
const tslib_1 = require("tslib");
const Link_1 = require("@prismicio/types-internal/lib/documents/widgets/nestable/Link");
const t = tslib_1.__importStar(require("io-ts"));
const mapOutput_1 = require("io-ts-types/lib/mapOutput");
const AnyLink_1 = require("./AnyLink");
const DocumentLink_1 = require("./DocumentLink");
const MediaLink_1 = require("./MediaLink");
const WebLink_1 = require("./WebLink");
const LinkDefConfig = t.union([MediaLink_1.MediaLinkDef, DocumentLink_1.DocumentLinkDef, WebLink_1.WebLinkDef, AnyLink_1.AnyLinkDef]);
exports.LinkDefType = "LinkDefType";
exports.LinkDef = (0, mapOutput_1.mapOutput)(t.strict({
    link: t.string,
    config: LinkDefConfig,
}), (o) => ({ name: o.link, __TYPE__: exports.LinkDefType, config: o.config }));
function fromLink(key, model) {
    const config = (() => {
        var _a;
        switch ((_a = model.config) === null || _a === void 0 ? void 0 : _a.select) {
            case "document":
                return (0, DocumentLink_1.fromDocumentLink)(model);
            case "media":
                return (0, MediaLink_1.fromMediaLink)(model);
            case "web":
                return (0, WebLink_1.fromWebLink)(model);
            default:
                return (0, AnyLink_1.fromAnyLink)(model);
        }
    })();
    return { link: key, config };
}
exports.fromLink = fromLink;
function defaultLinkDef(type) {
    const config = (() => {
        switch (type) {
            case Link_1.LinksTypes.DocumentLink:
                return (0, DocumentLink_1.defaultDocumentLinkDef)();
            case Link_1.LinksTypes.ImageLink:
            case Link_1.LinksTypes.FileLink:
                return (0, MediaLink_1.defaultMediaLinkDef)();
            case Link_1.LinksTypes.ExternalLink:
                return (0, WebLink_1.defaultWebLinkDef)();
        }
    })();
    return {
        name: "",
        config,
        __TYPE__: "LinkDefType",
    };
}
exports.defaultLinkDef = defaultLinkDef;
const isMediaLinkDef = (linkDef) => linkDef.config.__TYPE__ === MediaLink_1.MediaLinkDefType;
exports.isMediaLinkDef = isMediaLinkDef;
const isAnyLinkDef = (linkDef) => linkDef.config.__TYPE__ === AnyLink_1.AnyLinkDefType;
exports.isAnyLinkDef = isAnyLinkDef;
const isDocumentLinkDef = (linkDef) => linkDef.config.__TYPE__ === DocumentLink_1.DocumentLinkDefType;
exports.isDocumentLinkDef = isDocumentLinkDef;
const isWebLinkDef = (linkDef) => linkDef.config.__TYPE__ === WebLink_1.WebLinkDefType;
exports.isWebLinkDef = isWebLinkDef;
tslib_1.__exportStar(require("./AnyLink"), exports);
tslib_1.__exportStar(require("./DocumentLink"), exports);
tslib_1.__exportStar(require("./MediaLink"), exports);
tslib_1.__exportStar(require("./WebLink"), exports);
