// @prismicio/api-renderer/lib/models/ApiModel/slices/CompositeSlice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fromCompositeSlice = exports.CompositeSliceDef = exports.CompositeSliceDefType = void 0;
const tslib_1 = require("tslib");
const widgets_1 = require("@prismicio/types-internal/lib/customtypes/widgets");
const function_1 = require("@prismicio/types-internal/lib/validators/function");
const t = tslib_1.__importStar(require("io-ts"));
const Group_1 = require("../Group");
const SimpleField_1 = require("../SimpleField");
exports.CompositeSliceDefType = "compositeSliceDef";
exports.CompositeSliceDef = (0, function_1.addType)(t.type({
    name: t.string,
    primary: t.array(SimpleField_1.SimpleFieldDef),
    items: Group_1.GroupDef,
}), exports.CompositeSliceDefType);
function fromCompositeSlice(key, model) {
    return {
        name: key,
        primary: Object.entries(model["non-repeat"] || {}).map(([k, w]) => (0, SimpleField_1.fromSimpleField)(k, w)),
        items: (0, Group_1.fromGroup)(key, {
            type: widgets_1.WidgetTypes.Group,
            config: { fields: model.repeat || {} },
        }),
    };
}
exports.fromCompositeSlice = fromCompositeSlice;
