// @prismicio/api-renderer/lib/models/ApiModel/slices/SharedSlice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fromSharedSlice = exports.SharedSliceDef = exports.SharedSliceDefType = exports.VariationDef = exports.VariationDefType = void 0;
const tslib_1 = require("tslib");
const function_1 = require("@prismicio/types-internal/lib/validators/function");
const t = tslib_1.__importStar(require("io-ts"));
const SimpleField_1 = require("../SimpleField");
exports.VariationDefType = "variationDef";
exports.VariationDef = (0, function_1.addType)(t.strict({
    id: t.string,
    primary: t.array(SimpleField_1.SimpleFieldDef),
    items: t.array(SimpleField_1.SimpleFieldDef),
    version: t.string,
}), exports.VariationDefType);
exports.SharedSliceDefType = "sharedSliceDef";
exports.SharedSliceDef = (0, function_1.addType)(t.strict({
    name: t.string,
    variations: t.array(exports.VariationDef),
}), exports.SharedSliceDefType);
function fromSharedSlice(key, model) {
    return {
        name: key,
        variations: model.variations.map((v) => ({
            id: v.id,
            primary: Object.entries(v.primary || {}).map(([k, w]) => (0, SimpleField_1.fromSimpleField)(k, w)),
            items: Object.entries(v.items || {}).map(([k, w]) => (0, SimpleField_1.fromSimpleField)(k, w)),
            version: v.version,
        })),
    };
}
exports.fromSharedSlice = fromSharedSlice;
