// @prismicio/types-internal/lib/documents/widgets/nestable/EmbedContent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmbedContentC = exports.EmbedContent = exports.EmbedContentType = void 0;
const tslib_1 = require("tslib");
const fp_ts_1 = require("fp-ts");
const function_1 = require("fp-ts/function");
const t = (0, tslib_1.__importStar)(require("io-ts"));
const validators_1 = require("../../../validators");
const function_2 = require("../../../validators/function");
exports.EmbedContentType = "EmbedContent";
exports.EmbedContent = new t.Type("EmbedContent", (u) => u.__TYPE__ === exports.EmbedContentType, (u) => {
    return (0, function_1.pipe)(exports.EmbedContentC.decode(u), fp_ts_1.either.map((s) => ({
        ...exports.EmbedContentC.encode(s),
        all: u,
    })));
}, (ec) => exports.EmbedContentC.encode(ec));
exports.EmbedContentC = (0, function_2.addType)(t.exact(t.intersection([
    t.type({
        embed_url: t.string,
        type: t.string,
    }),
    t.partial({
        version: t.union([t.string, t.number, t.null]),
        title: validators_1.StringOrNull,
        author_name: validators_1.StringOrNull,
        author_url: validators_1.StringOrNull,
        provider_name: validators_1.StringOrNull,
        provider_url: validators_1.StringOrNull,
        cache_age: t.union([t.string, t.number, t.null]),
        thumbnail_url: validators_1.StringOrNull,
        thumbnail_width: validators_1.NumberOrNull,
        thumbnail_height: validators_1.NumberOrNull,
        html: validators_1.StringOrNull,
    }),
])), exports.EmbedContentType);
