// @prismicio/types-internal/lib/documents/widgets/nestable/ImageContent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImageContent = exports.ImageContentType = exports.ImageContentView = void 0;
const tslib_1 = require("tslib");
const t = (0, tslib_1.__importStar)(require("io-ts"));
const validators_1 = require("../../../validators");
const function_1 = require("../../../validators/function");
const Origin = t.exact(t.type({
    id: t.string,
    url: t.string,
    width: t.number,
    height: t.number,
}));
exports.ImageContentView = t.exact(t.intersection([
    t.type({
        origin: Origin,
        width: t.number,
        height: t.number,
        edit: t.type({
            zoom: t.number,
            crop: t.type({
                x: t.number,
                y: t.number,
            }),
            background: t.string,
        }),
    }),
    t.partial({
        url: t.string,
        credits: validators_1.NonEmptyStringOrNull,
        alt: validators_1.NonEmptyStringOrNull,
        provider: validators_1.StringOrNull,
    }),
]));
exports.ImageContentType = "ImageContent";
exports.ImageContent = (0, function_1.addType)(t.intersection([
    exports.ImageContentView,
    t.partial({
        thumbnails: t.record(t.string, exports.ImageContentView),
    }),
]), exports.ImageContentType);
