// @prismicio/types-internal/lib/documents/widgets/nestable/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NestableTypes = exports.Links = exports.IntegrationFieldsContent = exports.ImageContentView = exports.ImageContent = exports.GeoPointContent = exports.FieldContent = exports.BooleanContent = void 0;
const tslib_1 = require("tslib");
const BooleanContent_1 = require("./BooleanContent");
const EmbedContent_1 = require("./EmbedContent");
const FieldContent_1 = require("./FieldContent");
const GeoPointContent_1 = require("./GeoPointContent");
const ImageContent_1 = require("./ImageContent");
const IntegrationFieldsContent_1 = require("./IntegrationFieldsContent");
const Link_1 = require("./Link");
const SeparatorContent_1 = require("./SeparatorContent");
const StructuredTextContent_1 = require("./StructuredTextContent");
var BooleanContent_2 = require("./BooleanContent");
Object.defineProperty(exports, "BooleanContent", { enumerable: true, get: function () { return (0, tslib_1.__importDefault)(BooleanContent_2).default; } });
(0, tslib_1.__exportStar)(require("./EmbedContent"), exports);
var FieldContent_2 = require("./FieldContent");
Object.defineProperty(exports, "FieldContent", { enumerable: true, get: function () { return (0, tslib_1.__importDefault)(FieldContent_2).default; } });
var GeoPointContent_2 = require("./GeoPointContent");
Object.defineProperty(exports, "GeoPointContent", { enumerable: true, get: function () { return GeoPointContent_2.GeoPointContent; } });
var ImageContent_2 = require("./ImageContent");
Object.defineProperty(exports, "ImageContent", { enumerable: true, get: function () { return ImageContent_2.ImageContent; } });
Object.defineProperty(exports, "ImageContentView", { enumerable: true, get: function () { return ImageContent_2.ImageContentView; } });
var IntegrationFieldsContent_2 = require("./IntegrationFieldsContent");
Object.defineProperty(exports, "IntegrationFieldsContent", { enumerable: true, get: function () { return (0, tslib_1.__importDefault)(IntegrationFieldsContent_2).default; } });
exports.Links = (0, tslib_1.__importStar)(require("./Link"));
(0, tslib_1.__exportStar)(require("./SeparatorContent"), exports);
(0, tslib_1.__exportStar)(require("./StructuredTextContent"), exports);
exports.NestableTypes = {
    Links: Link_1.LinksTypes,
    StructuredText: StructuredTextContent_1.StructuredTextContentType,
    Boolean: BooleanContent_1.BooleanContentType,
    Embed: EmbedContent_1.EmbedContentType,
    Field: FieldContent_1.FieldContentType,
    GeoPoint: GeoPointContent_1.GeoPointContentType,
    Image: ImageContent_1.ImageContentType,
    IntegrationFields: IntegrationFieldsContent_1.IntegrationFieldsContentType,
    Separator: SeparatorContent_1.SeparatorContentType,
};
